data_structures/lru_cache.cpp

```diff
@@ -0,0 +1,177 @@
+ * you to quickly identify which item hasn't been used for the longest amount of
+ * time. Implemented using Static Doubly Circular Linked List
+ * @author [Shubh Bansal](https://github.com/proRamLOGO)
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for io operations
+#include <unordered_map>  /// for hashmap
+
+/**
+ * @namespace data_structures
+ * @brief Algorithms with data structures
+ */
+namespace data_structures {
+/**
+ * @namespace lru_cache
+ * @brief Functions for [Least Recently Used
+ * Cache](https://www.geeksforgeeks.org/lru-cache-implementation/)
+ * implementation
+ */
+namespace lru_cache {
+/**
+ * Class node
+ * Class node to build up and implement doubly circular linked lists
+ * which will be used as LRU Cache DS.
+ */
+class node {
+ public:
+    int key;  ///< Variable to contain value of {key}
+    int val;  ///< Variable to contain value of {value} associated with {key}
+    node
+        *next;  //< Pointer to point on the node appearing next to current node.
+    node *prev;  //< Pointer to point on the node appearing previous to current
+                 //node.
+
+    /**
+     * Constructor Function
+     * Accepts 2 parameters key and value for all new instaces of class node
+     * created.
+     * @param key : Key
+     * @param val : Value
+     * @returns NOTHING
+     */
+    node(int key, int val) {
+        this->key = key;
+        this->val = val;
+        this->next = this->prev = nullptr;
+    }
+};
+
+/**
+ * Class LRUCache
+ * Main Class LRUCache.
+ * Attributes:-
+ * - node *root ///< Will contain LRU value
+ * - unordered_map< int , node* > ///< To keep track od addresses of all nodes
+ * involved in linked list. Methods:-
+ * - LRUCache(int): Constructor Function
+ * - VALUE_TYPE get(int): Function to query on the LRU
+ * - void put(int,int): Function to add/update a key in LRU.
+ */
+class LRUCache {
+ public:
+    node *root;
+    std::unordered_map<int, node *> addr;
+
+    /**
+     * Constructor Function
+     * Accepts one parameter, the physical size for LRU.
+     * Creates a doubly circular linked list containg {capacity} number of
+     * nodes.
+     * @param capacity the physical size for LRU as needed for particular
+     * instance.
+     * @returns NOTHING
+     */
+    explicit LRUCache(int capacity) {
+        this->root = new node(-1, -1);
+        node *temp = this->root;
+        for (int i = 1; i < capacity; ++i) {
+            temp->next = new node(-1, -1);
+            temp->next->prev = temp;
+            temp = temp->next;
+        }
+        temp->next = this->root;
+        this->root->prev = temp;
+
+        // CLL of size capacity initialized.
+    }
+
+    /**
+     * get Function
+     * Accepts one parameter, the key whose value needs to be returned.
+     * If key requested is not present return -1.
+     * @param key value for this key is to be returned.
+     * @returns NOTHING
```

Comment:
This function **does** return something.
Please document it well, make it understandable.